internal/bucket/lifecycle/lifecycle.go

```diff
@@ -309,99 +310,175 @@ func (o ObjectOpts) ExpiredObjectDeleteMarker() bool {
 	return o.DeleteMarker && o.NumVersions == 1
 }
 
-// ComputeAction returns the action to perform by evaluating all lifecycle rules
-// against the object name and its modification time.
-func (lc Lifecycle) ComputeAction(obj ObjectOpts) Action {
-	action := NoneAction
+type lifecycleEvent struct {
+	EventAction Action
+	RuleID      string
+	Due         time.Time
+}
+
+type lifecycleEvents []lifecycleEvent
+
+func (es lifecycleEvents) Len() int {
+	return len(es)
+}
+
+func (es lifecycleEvents) Swap(i, j int) {
+	es[i], es[j] = es[j], es[i]
+}
+
+func (es lifecycleEvents) Less(i, j int) bool {
+	if es[i].Due.Equal(es[j].Due) {
+		// Prefer Expiration over Transition for both current and noncurrent
+		// versions
+		switch es[i].EventAction {
+		case DeleteAction, DeleteVersionAction:
+			return true
+		}
+		switch es[j].EventAction {
+		case DeleteAction, DeleteVersionAction:
+			return false
+		}
+		return true
+	}
+
+	// Prefer earlier occurring event
+	return es[i].Due.Before(es[j].Due)
+}
+
+// eval returns the lifecycle event applicable at now. If now is the zero value of time.Time, it returns the upcoming lifecycle event.
+func (lc Lifecycle) eval(obj ObjectOpts, now time.Time) lifecycleEvent {
+	var events []lifecycleEvent
 	if obj.ModTime.IsZero() {
-		return action
+		return lifecycleEvent{}
 	}
+
+	// Handle expiry of restored object; NB Restored Objects have expiry set on
+	// them as part of RestoreObject API. They aren't governed by lifecycle
+	// rules.
+	if !obj.RestoreExpires.IsZero() && now.After(obj.RestoreExpires) {
```

Comment:
Probably not the best place to put this code, my guess is the expiration should be effective even after we remove the lifecycle document from the bucket, but we can revisit this in the future

Reply:
Good catch! yes, we should expire the restored object even if the original lifecycle rule has been removed or disabled.